app/models/manageiq/providers/inventory/persister/builder/shared.rb

```diff
@@ -72,9 +72,31 @@ def hardwares
     end
 
     def operating_systems
+      custom_save_block = lambda do |_ems, inventory_collection|
+        vms_and_templates_ids = inventory_collection.data.map { |os| os.vm_or_template&.id }.compact
```

Comment:
Instead of ` os.vm_or_template&.id` could you just do `os.vm_or_template_id`, or does the inventory_collection not store data like Rails does?

Reply:
Doesn't work like that, this is what the payload looks like:
```
inventory_collection.data.first.data
{:vm_or_template=>InventoryObject:('vm-1411', InventoryCollection:<VmOrTemplate>, blacklist: [parent, resource_pool]), :product_name=>"CentOS 4/5 or later (64-bit)"}
```